packages/protocol/contracts/identity/Random.sol

```diff
@@ -88,25 +105,31 @@ contract Random is IRandom, Ownable, Initializable {
    * @notice Add a value to the randomness history.
    * @param blockNumber Current block number.
    * @param randomness The new randomness added to history.
+   * @param isLastEpochBlock If current block is the last of its epoch.
    * @dev The calls to this function should be made so that on the next call, blockNumber will
    * be the previous one, incremented by one.
    */
-  function addRandomness(uint256 blockNumber, bytes32 randomness) internal {
+  function addRandomness(uint256 blockNumber, bytes32 randomness, bool isLastEpochBlock) internal {
     history[blockNumber] = randomness;
-    if (historySize == 0) {
-      historyFirst = block.number;
-      historySize = 1;
-    } else if (historySize > randomnessBlockRetentionWindow) {
-      delete history[historyFirst];
-      delete history[historyFirst + 1];
-      historyFirst += 2;
-      historySize--;
-    } else if (historySize == randomnessBlockRetentionWindow) {
-      delete history[historyFirst];
-      historyFirst++;
+    if (isLastEpochBlock) {
+      delete history[lastEpochBlock];
+      lastEpochBlock = block.number;
     } else {
-      // historySize < randomnessBlockRetentionWindow
-      historySize++;
+      if (historySize == 0) {
+        historyFirst = block.number;
+        historySize = 1;
+      } else if (historySize > randomnessBlockRetentionWindow) {
+        delete history[historyFirst];
```

Comment:
Looks like this would cause us to delete the randomness of the last epoch block

Reply:
Agreed. I was initially confused if this case would ever be reached, but now I'm thinking this case is to handle if we set `randomnessBlockRetentionWindow` to a new value less than `historySize`. I'll add another conditional